modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientClusterGroupImpl.java

```diff
@@ -75,21 +75,22 @@ class ClientClusterGroupImpl implements ClientClusterGroup {
      *
      */
     ClientClusterGroupImpl(ReliableChannel ch, ClientBinaryMarshaller marsh) {
-        this.ch = ch;
-
-        utils = new ClientUtils(marsh);
-
-        projectionFilters = ProjectionFilters.FULL_PROJECTION;
+        this(ch, new ClientUtils(marsh), ProjectionFilters.FULL_PROJECTION);
     }
 
     /**
      *
      */
-    private ClientClusterGroupImpl(ReliableChannel ch, ClientUtils utils,
-        ProjectionFilters projectionFilters) {
+    private ClientClusterGroupImpl(
+        ReliableChannel ch,
+        ClientUtils utils,
+        ProjectionFilters projectionFilters
+    ) {
         this.ch = ch;
         this.utils = utils;
         this.projectionFilters = projectionFilters;
+
+        ch.addChannelFailListener(this::invalidateCachedData);
```

Comment:
There is memory leak. Every time cluster group is created - reference to this group is stored to the channel and this group can't be garbage collected until client closed.